---
title: "BCB420: Assignment 2"
subtitle: "Differential Gene Expression and Preliminary Over-Representation Analysis of Neutrophils Affected by Triple-Negative Breast Cancer (TNBC)"
author: "Katarina Vucic"
date: "03/11/2025"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    theme: flatly
    highlight: tango
    fig_caption: true
    df_print: paged 
    code_folding: hide
bibliography: A2_references.bib
nocite: '@*'
link-citations: true
---

# Introduction

While cancer is one of the most studied diseases in the world, the immune landscape of cancer patients is largely uncharacterized. Previous research has demonstrated that tumours can affect the immune system's response, leading to variations in disease progression and treatment outcomes [@blomberg2018; @mcallister2014]. Neutrophils are a type of white blood cell that play a critical role in the body's immune response [@nci2019]. They respond to conditions in the body by releasing enzymes to destroy microorganisms [@nci2019]. Neutrophils also help to activate other immune cells, making them vital to the immune system [@nci2019]. 

Triple-Negative Breast Cancer (TNBC) is a particularly aggressive breast cancer subtype with a high mortality rate [@acs2023]. To investigate immune response in TNBC patients, Bakker et. al performed a bulk RNA-seq experiment designed to identify the differences in the expression of genes in the neutrophils of TNBC patients and healthy donors [@bakker2025]. In this analysis, we continue the work of Bakker et. al by examining their bulk RNA-sequencing results to investigate the expression patterns of genes in TNBC patients.

In previous work, we have taken this dataset and cleaned it for further analysis by removing low read counts and duplicated data before performing Trimmed Mean of M-values (TMM) normalization. In the process of cleaning and normalizing the data, we were left with `3690` genes in our bulk RNA-sequencing experiment. This dataset includes the neutrophils of 7 mTNBC patients in the test set and 7 healthy donors in the control group. All Ensembl IDs are mapped to HGNC symbols, and the counts for duplicate Ensembl IDs for the same HGNC symbol were averaged. The quality checking showed that the data is clean with no missing or non-numeric values. To normalize the data, we used TMM normalization to reduce technical noise. Using the the MDS plot, we see that our samples group nicely by disease state.

Now, we will perform differential gene expression using `edgeR`, and conduct an over-representation analysis. 

TODO: flesh out intro and add more detail. maybe include some figures.

---

## Data Import and Cleaning

We will begin by importing our normalized TNBC data. 

```{r}
if (!requireNamespace("ggrepel", quietly = TRUE)) {
  install.packages("ggrepel")
}

library(edgeR)
library(ComplexHeatmap)
library(circlize)
library(ggrepel)
library(ggplot2)
library(gprofiler2)
```

```{r}
# load in data from csv file
normalized_tnbc_data <- read.csv("normalized_tnbc_data.csv")

# reset index and remove the row
rownames(normalized_tnbc_data) <- normalized_tnbc_data$X
normalized_tnbc_data$X <- NULL
```
*This code was inspired by the response by Matthew Lundberg to this Stack Overflow question [].*
(https://stackoverflow.com/questions/20643166/set-a-data-frame-column-as-the-index-of-r-data-frame-object)

As a reminder, our dataframe contains the nromalized read counts of 3690 genes for 14 samples.

```{r}
normalized_tnbc_data
```

Let's take a look at the distribution of the data to confirm that it is in fact normalized.

```{r, message=FALSE, warning=FALSE}
plot_count_distribution <- function(count_data, title = "RNASeq Samples") {
  data2plot <- log2(count_data)
  boxplot(data2plot, xlab = "Samples", ylab = "log2 TPM",
    las = 2, cex = 0.5, cex.lab = 0.5,
    cex.axis = 0.5, main = "RNASeq Samples")
  #draw the median on each box plot
  abline(h = median(apply(data2plot, 2, median)),
    col = "green", lwd = 0.6, lty = "dashed")
  mtext("Boxplot of TNBC RNASeq data for each sample", side = 1, line = 4, cex = 0.75)
  
  counts_density <- apply(log2(count_data), 2, density)
    #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(counts_density)) {
      xlim <- range(c(xlim, counts_density[[i]]$x));
      ylim <- range(c(ylim, counts_density[[i]]$y))
    }
    cols <- rainbow(length(counts_density))
    ltys <- rep(1, length(counts_density))
    #plot the first density plot to initialize the plot
    plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
      ylab="Smoothing density of log2-CPM",
      main=paste(title, "- Density Plot"), cex.lab = 0.85)
    #plot each line
    for (i in 1:length(counts_density))
      lines(counts_density[[i]], col=cols[i], lty=ltys[i])
    #create legend
    legend("topright", colnames(data2plot),
      col=cols, lty=ltys, cex=0.75,
      border ="blue", text.col = "green4",
      merge = TRUE, bg = "gray90")
    mtext("Desnity plot of TNBC RNASeq data for each sample.", side = 1, line = 4, cex = 0.75)
}

plot_count_distribution(normalized_tnbc_data, title = "Filtered TNBC Data")
```

*This code was modified from the lectures for BCB420: Computational Systems Biology [@isserlin2025].*

Since our data has aligned means across samples, with the distribution resembling a normal distribution, we can conclude that our data is normalized, as expected. 

Lets load in the sample dataframe next.

```{r}
# load in data from csv file
sample_info_data <- read.csv("sample_info_data.csv")

# reset index and remove the row
rownames(sample_info_data) <- sample_info_data$X
sample_info_data$X <- NULL
```
*This code was inspired by the response by Matthew Lundberg to this Stack Overflow question [].*
(https://stackoverflow.com/questions/20643166/set-a-data-frame-column-as-the-index-of-r-data-frame-object)

We have biological sample information for all 14 samples.

```{r}
sample_info_data
```

Now that we are sure our data is clean and normalized, we can begin our analysis.

## Data Separation

In order to determine what category we should use to calculate differential expression, we will look at how our data separates according to each classification that we have data for. We will use a **Multidimensional Scaling (MDS)** plot to help us assess the similarities between samples based on their expression profiles [@hout2013]. This will show us how well the samples are grouped by their biological characteristics, such as disease state and clinical trial [@hout2013].

### Data Separation by Disease State

We will begin by assessing how well the samples separate by disease state.

```{r, message=FALSE, warning=FALSE}
tnbc_dge_list <- edgeR::DGEList(counts=normalized_tnbc_data, group=sample_info_data$diseaseState)
tnbc_norm_factors <- calcNormFactors(tnbc_dge_list)

limma::plotMDS(tnbc_norm_factors, labels=NULL, pch = 1,
               col = c("darkgreen","blue")[factor(sample_info_data$diseaseState)])

legend("topright", 
       legend=levels(factor(sample_info_data$diseaseState)),
  pch=c(1), col= c("darkgreen","blue"),title="Class",
  bty = 'n', cex = 0.75)

mtext("MDS plot of TNBC RNASeq data coloured by disease state.", side = 1, line = 4, cex = 0.75)
```

There is a very clear separation of samples by disease state, indicating that this is a very predictive feature for our model. This is a good sign, considering our goal is characterizing the difference in disease state between healthy and TNBC samples!


### Data Separation by Clinical Trial

Let's also take a look at how samples cluster according to clinical trial.

```{r, message=FALSE, warning=FALSE}
tnbc_dge_list_trial <- edgeR::DGEList(counts=normalized_tnbc_data, group=sample_info_data$trial)
tnbc_norm_factors_trial <- calcNormFactors(tnbc_dge_list)

limma::plotMDS(tnbc_norm_factors_trial, labels=NULL, pch = 1,
               col = c("darkgreen","blue", "purple")[factor(sample_info_data$trial)])

legend("topright", 
       legend=levels(factor(sample_info_data$trial)),
  pch=c(1), col= c("darkgreen","blue", "purple"),title="Class",
  bty = 'n', cex = 0.75)

mtext("MDS plot of TNBC RNASeq data coloured by clinical trial", side = 1, line = 4, cex = 0.75)
```

There is no distinct separation between TNB and Tonic samples, which is good! It means that trial does not add any obvious technical variability that may get in the way when looking for biological variability.

Ideally, we would have liked to look at how the samples separate in response to cancer stage and chemotherapy level, however, the authors did not include that information in the sample information. We will stick with disease state as our main factor for the model we will use for differential expression.

# Differential Gene Expression

TODO explain DE

We will use the the `edgeR` package with the `glmQLFit` function to calculate differential expression. We will begin by creating a model to separate by disease state and then fitting the data to our model.

```{r}
# set up model based off disease state
model_design_pattern <- model.matrix(~ sample_info_data$diseaseState)

# calculate dispersion and fit it to the model
tnbc_dispersion <- estimateDisp(tnbc_dge_list, model_design_pattern)
tnbc_fit <- glmQLFit(tnbc_dispersion, model_design_pattern)
```

Now, we are ready to run our differential expression test.
```{r}
qlf_disease_state <- glmQLFTest(tnbc_fit, coef='sample_info_data$diseaseStateTriple-Negative Breast Cancer')
qlf_output_hits <- topTags(qlf_disease_state, sort.by = "PValue", n = nrow(normalized_tnbc_data))
```

Let's take a look at our results. Specifically, we will look at how many genes were significantly differentially expressed with a p-value < 0.05. This is the standard p-value threshold for statistical significance.

```{r}
results_table <- qlf_output_hits$table
results_table$gene <- rownames(results_table)

print(length(results_table$gene[results_table$PValue < 0.05]))
```

There are 572 significantly differentially expressed genes in this dataset.

## Multiple Hypothesis Testing

TODO add section explaining what it is 

### Benjamini-Hochberg FDR Correction

We will apply Benjamini-Hochberg (BH) correction for multiple hypothesis testing. The BH method controls the False Discovery Rate (FDR) instead of the Family-Wise Error Rate (FWER) like other methods, making it less-stringent and more beneficial for discovery (https://link.springer.com/referenceworkentry/10.1007/978-1-4419-9863-7_1215). We will use the FDR values provided by edgeR.

Additionally, after reading several papers and discussion forums, we will go with a FDR cutoff of 0.1. The general consensus in the scientific discovery community is that an FDR value < 0.1 is acceptable in the case of exploratory studies, where we are prioritizing discovery over stringent control of false positives (https://www.statsig.com/blog/understanding-significance-levels-a-key-to-accurate-data-analysis, https://groups.google.com/g/trinityrnaseq-users/c/T9VBfIAMbjU?pli=1). Other packages also make use of this cutoff, for example, the standard FDR cutoff in DeSeq2 is 0.1 (https://bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf).

```{r}
print(results_table$gene[results_table$FDR <= 0.1 & results_table$PValue < 0.05])
```

We are left with 13 genes that pass our BH correction with FDR < 0.01.

### Visualizing Differentially Expressed Genes

#### Volcano Plot

Let's take a look at the amount of differentially expressed genes in our dataset. The genes that pass correction are labelled and have a star point. Those coloured in red are UP regulated and those in blue are DOWN regulated, with a p-value of 0.05 and a log fold-change cutoff of 1.1. The choice of logFC inspired from this forum post https://support.bioconductor.org/p/98367/.

```{r}
# a column to categorize whether something is differentially expressed or not
results_table$diffexpressed <- "NO"
# if logFC > 1.1 and p-value < 0.05, set as "UP" 
results_table$diffexpressed[results_table$logFC > 1.1 & results_table$PValue < 0.05] <- "UP"
# if logFC < -1.1 and p-value < 0.05, set as "DOWN" 
results_table$diffexpressed[results_table$logFC < -1.1 & results_table$PValue < 0.05] <- "DOWN"

# add a column to label the differentially expressed genes that pass correction
results_table$label <- NA
results_table$label[results_table$FDR <= 0.1] <- results_table$gene[results_table$FDR <= 0.1]

# plot adding up all layers we have seen so far
ggplot(data=results_table, aes(x=logFC, y=-log10(PValue), col=diffexpressed, label=label)) +
        geom_point() + 
  geom_point(data=subset(results_table, !is.na(label)), shape=8, size=3) +  # star markers for labeled genes
        theme_minimal() +
        geom_text_repel() +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_vline(xintercept=c(-1.1, 1.1), col="red") +
        geom_hline(yintercept=-log10(0.05), col="red")

```

code inspred from https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html.

We have an almost even split across our target genes, with 6 genes down-regulated and 7 genes up-regulated.

#### Heatmap

Let's take a look at how the 572 significantly expressed differential genes separate according to disease state using a heatmap. Genes that are up-regulated in a sample with show up red on the heatmap, and those down-regulated will show up as blue. 

We will begin by row-normalizing our heatmap data so that the average expression level in each row is centred around 0. This way, if the expression for a gene is higher than average in one sample it will be red, and if its lower than average it will be blue.  

```{r}
# take out the top-hits with p-value < 0.05
top_hits <- rownames(qlf_output_hits$table)[qlf_output_hits$table$PValue<0.05]

# index only the top hits and row-normalize the data
heatmap_matrix_tophits <- t(scale(t(normalized_tnbc_data[which(rownames(normalized_tnbc_data) %in% top_hits),])))

# set up colors for the heatmap
heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0, max(heatmap_matrix_tophits)), 
                         c("blue", "white", "red"))


# create color mapping for disease state
unique_tnbc <- unique(sample_info_data$diseaseState)
unique_tnbc_colors <- rainbow(n = length(unique_tnbc))
names(unique_tnbc_colors) <- unique_tnbc

# set up annotation by disease state
ha_pat <- HeatmapAnnotation(df = data.frame(tnbc = sample_info_data$diseaseState),
                            col = list(tnbc = unique_tnbc_colors), show_legend = TRUE)

# plot heatmap
tnbc_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
                        top_annotation = ha_pat, # set up heatmap annotation
                        column_split = sample_info_data$diseaseState, # cluster by disease state
                        cluster_rows = TRUE, cluster_columns = TRUE, # cluster rows and columns
                        show_row_dend =TRUE, show_column_dend = TRUE,  # show row and column dendrograms
                        col=heatmap_col, # set up colors
                        show_column_names = FALSE, show_row_names = FALSE, # dont show row or col names
                        show_heatmap_legend = TRUE, # show legend
                        column_title = ("Top hits Healthy vs. TNBC"))

tnbc_heatmap
```

We can see that the genes in the top third of our heatmap appears to be lowly ezpressed in TNBC samples and the bottom two-thirds seem to be lowly expressed in healthy samples. We will return to this observation later.


# Over-Representation Analysis

We will use gprofiler to perform thresholded over-representation analysis (add info from lec).

Let's begin by creating our upregulated and downregulated gene sets. We declare a gene as significantly upregulated if it has a p-value < 0.05 and a log fold change > 1.1 (cite paper from before). We declare our significantly downregulated genes as those with a p-value < 0.05 and a log fold change < -1.1.

```{r}
upregulated_genes <- results_table$gene[which(results_table$PValue < 0.05 & results_table$logFC > 1.1)]

downregulated_genes <- results_table$gene[which(results_table$PValue < 0.05 & results_table$logFC < -1.1)]

combined_genes <- results_table$gene[which(results_table$PValue < 0.05 & (results_table$logFC > 1.1 | results_table$logFC < -1.1))]

top_genes <- results_table$gene[which(results_table$PValue < 0.05 & results_table$FDR < 0.1)]

print(length(upregulated_genes))
print(length(downregulated_genes))
```


We have 78 differentially expressed upregulated genes and 155 differentially expressed downregulated genes.
We will use GO: Biological Process. This is the most common dataset for enrichment analysis, since it helps us understand the biological functions and pathways that are significantly enriched in our gene set (https://pmc.ncbi.nlm.nih.gov/articles/PMC6607905/#:~:text=Gene%20Ontology%20(GO)57:%20GO%20provides%20a%20hierarchically,based%20on%20these%20terms%20for%20multiple%20species.&text=Most%20enrichment%20analysis%20methods%20make%20unrealistic%20assumptions,independence%20among%20genes%20as%20well%20as%20pathways.).

## Overall Differential Expression Analysis

We will begin by performing differential expression on the full list.
```{r}
query_set <- combined_genes

gprofiler_results <- gost(query = query_set ,
                          significant=FALSE,
                          ordered_query = FALSE,
                          exclude_iea=TRUE,
                          correction_method = "fdr",
                          organism = "hsapiens",
                          source = c("GO:BP"))

combined_results <- as.data.frame(gprofiler_results$result)
print(combined_results)
```

We do not get any significant results when we include both sets. 

## Up-Regulated Differential Expression Analysis

Let's see what happens when we perform our analysis on the upregulated and downregulated genes separately. We will begin with upregulated genes

```{r}
query_set <- upregulated_genes

gprofiler_results <- gost(query = query_set ,
                          significant=TRUE,
                          ordered_query = FALSE,
                          exclude_iea=TRUE,
                          correction_method = "fdr",
                          organism = "hsapiens",
                          source = c("GO:BP"))

upregulated_results <- as.data.frame(gprofiler_results$result)
print(upregulated_results)
```

For upregulated genes, the significant hits are defense response to fungus, response to fungus, positive regulation of blood coagulation, positive regulation of hemostasis, and positive regulation of coagulation. Much to think about.

## Down-Regulated Differential Expression Analysis

Now let's try with the down-regulated genes

```{r}
query_set <- downregulated_genes

gprofiler_results <- gost(query = query_set ,
                          significant=TRUE,
                          ordered_query = FALSE,
                          exclude_iea=TRUE,
                          correction_method = "fdr",
                          organism = "hsapiens",
                          source = c("GO:BP"))

downregulated_results <- as.data.frame(gprofiler_results$result)
print(downregulated_results)
```


The most common pathways here are macromolecule metabolic process	
mitochondrial translation	
mitochondrial gene expression	
eosinophil migration	
chemokine-mediated signaling pathway
eosinophil chemotaxis	
macromolecule biosynthetic process	
response to chemokine	
cellular response to chemokine	
gene expression

## FDR-Corrected Gene Differential Expression Analysis

```{r}
query_set <- top_genes

gprofiler_results <- gost(query = query_set ,
                          significant=FALSE,
                          ordered_query = FALSE,
                          exclude_iea=TRUE,
                          correction_method = "fdr",
                          organism = "hsapiens",
                          source = c("GO:BP"))

top_results <- as.data.frame(gprofiler_results$result)
print(top_results)
```

Our top expressed genes all belong to different groups, so there is no significant pathway between them.

TODO: code inspired by https://risserlin.github.io/CBW_pathways_workshop_R_notebooks/run-gprofiler-from-r.html for all gprofiler runs.

