---
title: "BCB420: Assignment 2"
subtitle: "Differential Gene Expression and Preliminary Over-Representation Analysis of Neutrophils Affected by Triple-Negative Breast Cancer (TNBC)"
author: "Katarina Vucic"
date: "03/11/2025"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    theme: flatly
    highlight: tango
    fig_caption: true
    df_print: paged 
    code_folding: hide
bibliography: A2_references.bib
nocite: '@*'
link-citations: true
---

# Introduction

While cancer is one of the most studied diseases in the world, the immune landscape of cancer patients is largely uncharacterized. Previous research has demonstrated that tumours can affect the immune system's response, leading to variations in disease progression and treatment outcomes [@blomberg2018; @mcallister2014]. Neutrophils are a type of white blood cell that play a critical role in the body's immune response [@nci2019]. They respond to conditions in the body by releasing enzymes to destroy microorganisms [@nci2019]. Neutrophils also help to activate other immune cells, making them vital to the immune system [@nci2019]. 

Triple-Negative Breast Cancer (TNBC) is a particularly aggressive breast cancer subtype with a high mortality rate [@acs2023]. To investigate immune response in TNBC patients, Bakker et. al performed a bulk RNA-seq experiment designed to identify the differences in the expression of genes in the neutrophils of TNBC patients and healthy donors [@bakker2025]. In this analysis, we continue the work of Bakker et. al by examining their bulk RNA-sequencing results to investigate the expression patterns of genes in TNBC patients.

In previous work, we have taken this dataset and cleaned it for further analysis by removing low read counts and duplicated data before performing Trimmed Mean of M-values (TMM) normalization. In the process of cleaning and normalizing the data, we were left with `3690` genes in our bulk RNA-sequencing experiment. This dataset includes the neutrophils of 7 mTNBC patients in the test set and 7 healthy donors in the control group. All Ensembl IDs are mapped to HGNC symbols, and the counts for duplicate Ensembl IDs for the same HGNC symbol were averaged. The quality checking showed that the data is clean with no missing or non-numeric values. To normalize the data, we used TMM normalization to reduce technical noise. Using the the MDS plot, we see that our samples group nicely by disease state.

Now, we will perform differential gene expression using `edgeR`, and conduct an over-representation analysis. 

#TODO: flesh out intro and add more detail. maybe include some figures.

---