---
title: "BCB420: Assignment 3"
subtitle: "Dataset Pathway and Network Analysis of Neutrophils Affected by Triple-Negative Breast Cancer (TNBC)"
author: "Katarina Vucic"
date: "04/01/2025"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: true
    theme: default
    highlight: tango
    fig_caption: true
    code_folding: hide
bibliography: A3_references.bib
nocite: '@*'
link-citations: true
---

# Introduction

Triple-Negative Breast Cancer (TNBC) is a particularly aggressive subtype of breast cancer, known for its high mortality rate [@americancancersociety_2023]. Although cancer research continues to evolve rapidly, the immune landscape of individuals affected by TNBC remains largely uncharacterized. Previous studies have demonstrated that tumors can alter immune system responses, resulting in differences in disease progression and treatment outcomes [@blomberg_2018; @mcallister_2014]. Neutrophils, a type of white blood cell, are crucial for immune defense, as they release enzymes that break down microorganisms and activate other immune cells, making them a vital component of the immune response [@nationalcancerinstitute_2019].

To investigate immune responses in TNBC patients, Bakker et al. performed a bulk RNA-seq experiment to analyze differential gene expression in neutrophils from both TNBC patients and healthy donors [@bakker_2025]. Expanding on their findings, we re-examine their bulk RNA-sequencing data to explore gene expression patterns in TNBC patients. Their dataset has been made publicly available in the Gene Expression Omnibus (GEO) under accession number GSE264108 [@barrett_2012].

In our previous work, we processed this dataset by removing low read counts and duplicated data before applying Trimmed Mean of M-values (TMM) normalization. After cleaning and normalizing the data, our bulk RNA-sequencing experiment retained 3690 genes. The dataset consists of neutrophils from 7 mTNBC patients in the test group and 7 healthy donors in the control group.

Furthermore, we performed a differential gene expression and over-representation analysis on the gene set obtained from Bakker et al.â€™s study. Our findings indicate that neutrophils in TNBC patients are enriched for pathways related to cell migration and neutrophil granule proteins, aligning with the results of their research. Additionally, we observed a down-regulation of mitochondrial translation proteins in TNBC samples, a condition that has been linked to breast cancer in previous studies [@koc_2022; @wang_2021].

For this analysis, we will conduct enrichment analysis on our dataset, to identify pathways that are enriched for up or down regulation in TNBC samples.

---

## Data Import and Cleaning

#TODO: fix imports ->

We begin by importing relevant packages for our analysis. We will be using edgeR [@edger] for differential expression; ComplexHeatmap [@complexheatmap], circlize [@circlize], ggrepel [@ggrepel], and ggplot2 [@ggplot2] for plotting; dplyr [@dplyr_2020] for data manipulation; kable extra [@kableextra_2021] for improved table styling; and gprofiler2 [@gprofiler2] for over-representation analysis.

```{r, message=FALSE, warning=FALSE, error=FALSE, results='hide'}
library(ggplot2)
library(knitr)
library(kableExtra)
library(dplyr)
```


#TODO: decide on A1/A2 append or not ->

```{r, child="A1_KatarinaVucic.Rmd"}
```

```{r, child="A2_KatarinaVucic.Rmd"}
```


First, we will import our normalized TNBC data. 

```{r, message=FALSE, warning=FALSE}
# load in data from csv file
normalized_tnbc_data <- read.csv("A2_normalized_tnbc_data.csv")

# reset index and remove the row
rownames(normalized_tnbc_data) <- normalized_tnbc_data$X
normalized_tnbc_data$X <- NULL

# display the data
kable(normalized_tnbc_data, format = "html", caption = "  Table 1: Normalized read counts for TNBC vs Healthy Neutrophils") %>%
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% # striped rows for readability, constrain width
  scroll_box(height = "400px") # set the height of the box so it won't be too big
```

*This code was inspired by the response by Matthew Lundberg to this Stack Overflow question [@brno792_2013], and the kable display was inspired by [@stevec_2020] and [@tierney_2020].*

Let's take a look at the distribution of the data to confirm that it is normalized.

```{r, message=FALSE, warning=FALSE}
data2plot <- log2(normalized_tnbc_data)
boxplot(data2plot, xlab = "Samples", ylab = "log2 TPM",
  las = 2, cex = 0.5, cex.lab = 0.5,
  cex.axis = 0.5, main = "Boxplot of RNASeq Sample Distributions")
#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)),
  col = "green", lwd = 0.6, lty = "dashed")
```

```{r, message=FALSE, warning=FALSE, fig.cap="Figure 1: Distribution of the normalized gene read counts for healthy and TNBC samples. A) A Boxplot of the distribution of read counts for each sample, with a centre line showing the mean. B) An overlaid line graph showing the density function for each sample. This code was modified from the lectures for BCB420: Computational Systems Biology [@isserlin_2025]."}
counts_density <- apply(log2(normalized_tnbc_data), 2, density)
#calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x));
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))
#plot the first density plot to initialize the plot
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
  ylab="Smoothing density of log2-CPM",
  main=paste("Density Plot of RNASeq Sample Distributions"), cex.lab = 0.85)
#plot each line
for (i in 1:length(counts_density))
  lines(counts_density[[i]], col=cols[i], lty=ltys[i])
#create legend
legend("topright", colnames(data2plot),
  col=cols, lty=ltys, cex=0.75,
  border ="blue", text.col = "green4",
  merge = TRUE, bg = "gray90")
```

Since our data has aligned means across samples, with the distribution resembling a normal distribution, we can conclude that our data is normalized, as expected. 

Let's load in the sample dataframe next.

```{r, message=FALSE, warning=FALSE}
# load in data from csv file
sample_info_data <- read.csv("A2_sample_info_data.csv")

# reset index and remove the row
rownames(sample_info_data) <- sample_info_data$X
sample_info_data$X <- NULL
```

*This code was inspired by the response by Matthew Lundberg to this Stack Overflow question [@brno792_2013].*

We have biological sample information for all 14 samples, with 7 from healthy donors and 7 from TNBC donors. Additionally, we have information about which clinical trial each TNBC sample is from, Tonic or TNB.

```{r, message=FALSE, warning=FALSE}
# display the data
kable(sample_info_data, format = "html", caption = "Table 2: Sample information for TNBC and healthy neutrophils.") %>%
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% # striped rows for readability, constrain width
  scroll_box(height = "400px") # set the height of the box so it won't be too big
```

Now that we are sure our data is clean and normalized, we can begin our analysis.

---

# Non-thresholded Gene Set Enrichment Analysis

---

## Results

---

## Enrichment Map

---

## Theme Network

---

## Additionl Analysis

---

# Discussion

---

# Conclusion

---


# Question Key

The following contains a guide to the discussion questions for A3. Clicking on the question links back to the section where the analysis was performed.

**Non-thresholded Gene Set Enrichment Analysis**  

[**What method did you use? What gene sets did you use? Make sure to specify versions and cite your methods. Summarize your enrichment results.**](#)  

[**Summarize your enrichment results.**](#) 

[**How do these results compare to the results from the thresholded analysis in Assignment #2? Compare qualitatively. Is this a straightforward comparison? Why or why not?**](#)  


**Visualizing Your Gene Set Enrichment Analysis in Cytoscape**  

[**How many nodes and how many edges are in the resulting map? What thresholds were used to create this map? Make sure to record all thresholds. Include a screenshot of your network prior to manual layout.**](#)  

[**What parameters did you use to annotate the network? If you are using the default parameters, make sure to list them as well.**](#)  

[**How did you make a publication-ready figure? Include this figure with proper legends in your notebook.**](#)  

[**What are the major themes present in the collapsed network? Do they fit with the model? Are there any novel pathways or themes?**](#)  


**Interpretation and Detailed View of Results**  

[**Do the enrichment results support conclusions or mechanisms discussed in the original paper? How do these results differ from the results you got from Assignment #2 thresholded methods?**](#)  

[**Can you find evidence (e.g., publications) to support some of the results that you see? How does this evidence support your result?**](#)  

[**What additional analyses can you add?**](#)  

---

# References



